pkg/sql/sem/builtins/builtinsregistry/builtins_registry.go

```diff
@@ -20,16 +20,40 @@ import "github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 
 var registry = map[string]definition{}
 
+// Subscription is a hook to be called once on all registered builtins.
+type Subscription func(name string, props *tree.FunctionProperties, overloads []tree.Overload)
+
+var subscriptions = make([]Subscription, 0)
+
 // Register registers a builtin. Intending to be called at init time, it panics
 // if a function of the same name has already been registered.
 func Register(name string, props *tree.FunctionProperties, overloads []tree.Overload) {
 	if _, exists := registry[name]; exists {
 		panic("duplicate builtin: " + name)
 	}
+	for i := range overloads {
+		var hook func()
+		overloads[i].OnTypeCheck = &hook
+	}
 	registry[name] = definition{
 		props:     props,
 		overloads: overloads,
 	}
+	for _, s := range subscriptions {
+		s(name, props, overloads)
+	}
+}
+
+// AddSubscription defines a hook to be called once on all registered builtins.
+// Subscriptions should be added at init() time, but are load-order independent:
+// if you add a subscription after a function is registered, it will immediately
+// be called on that function, while functions that are registered afterwards will
+// also trigger the hook.
```

Comment:
having a bit of a hard time seeing why we need this.
Can you explain what is the purpose of this?
Does it make sense to pull it into separate commit?